compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -554,6 +641,129 @@ impl<'tcx> TyCtxt<'tcx> {
     fn set_alloc_id_same_memory(self, id: AllocId, mem: &'tcx Allocation) {
         self.alloc_map.lock().alloc_map.insert_same(id, GlobalAlloc::Memory(mem));
     }
+
+    /// Store the result of a call to `eval_to_allocation_raw` in order to
+    /// allow deduplication.
+    #[instrument(skip(self), level = "debug")]
+    pub fn save_alloc_for_dedup(self, id: GlobalId<'tcx>, val: ConstDedupResult<ConstAlloc<'tcx>>) {
+        let dedup_const_map = self.dedup_const_map.lock();
+        dedup_const_map.insert_alloc(id, val);
+        debug!("dedup_const_map after insert: {:#?}", dedup_const_map);
+    }
+
+    /// Store the result of a call to `eval_to_const_value_raw` in order to deduplicate it.
+    #[instrument(skip(self), level = "debug")]
+    pub fn save_const_value_for_dedup(
+        self,
+        id: GlobalId<'tcx>,
+        val: ConstDedupResult<ConstValue<'tcx>>,
+    ) {
+        let dedup_const_map = self.dedup_const_map.lock();
+        dedup_const_map.insert_const_val(id, val);
+        debug!("dedup_const_map after insert: {:#?}", dedup_const_map);
+    }
+
+    /// Tries to deduplicate a call to `eval_to_allocation_raw`. If deduplication isn't
+    /// successful `eval_to_allocation_raw` query is executed.
+    #[instrument(skip(self, opt_span), level = "debug")]
+    pub fn dedup_eval_alloc_raw(
```

Comment:
these two functions are almost the same... do you think it would be convoluted to put their bodies into a separate generic function that takes two closures, one for the query and one for the dedup map lookup?